### DISCUSSION TITLE
How to setup nested entity resolver in async-graphql

### DISCUSSION THREAD

**Discussion by cchee** [async-graphql/async-graphql#1370] [Q&A] [answered]
First of all, this is a great library. The example codes are very helpful. I was able to quickly setup database schema along with graphql using this amazing library. But I am still a bit confused on how to setup nested entity query using async-graphql even after looking at the example code.

I am trying the get the customer query to be able to lookup a list of orders. I saw the example and tried to follow along to make it works with my database schema using sea-orm. 

Here is the customer.rs under entity.
`#[derive(Clone, Debug, Default, PartialEq, DeriveEntityModel, Serialize, Deserialize, SimpleObject)]`
`#[sea_orm(table_name = "customers", schema_name = "public")]`
`#[graphql(concrete(name = "Customer", params()))]`
`pub struct Model {`
`    #[sea_orm(primary_key)]`
`    #[graphql(shareable)]`
`    pub id: i32,`
`    #[sea_orm(indexed)]`
`    pub name: String,`
`}`
`// all the rest`
`:`
`:`

Here is the order.rs under entity.
`#[derive(Clone, Debug, Default, PartialEq, DeriveEntityModel, Serialize, Deserialize, SimpleObject)]`
`#[sea_orm(table_name = "orders", schema_name = "public")]`
`#[graphql(concrete(name = "Order", params()))]`
`pub struct Model {`
`    #[sea_orm(primary_key)]`
`    #[graphql(shareable)]`
`    pub id: i32,`
`    #[sea_orm(indexed)]`
`    pub customer_id: i32,`
`    #[sea_orm(indexed)]`
`    pub amount: f32,`
`}`
`// all the rest`
`:`
`:`

How do I change the code that makes use of async-graphql to support nested entity (order) in the following query?
`query {`
`   getCustomers {`
`      {`
`          id`
`          name`
`          orders {`
`             id`
`             amount`
`          }`
`      }`
`   }`
`}`

Apologize if this sounds like a dumb question, I am new to rust as well as async-graphql.

p.s. I managed to get federation working locally using rover dev, but I just couldn't figure out how to get the resolver for nested entity.

**Comment by cchee** [comment]
https://github.com/async-graphql/examples/blob/59cfedf21c1bee4c06a6345222a0d67ff34b20f1/federation/static-schema/federation-reviews/src/main.rs

Ok. Just stumble upon this one... Going to improvise and see if that helps...

**Comment by cchee** [comment]
I thought I could borrow and improvise the implementation for User-Reviews from the example. I added

`#[ComplexObject]`
`impl Customer {`
 `   async fn orders<'a>(&self, ctx: &'a Context<'_>) -> Vec<&'a Order> {`
 `       let orders = ctx.data_unchecked::<Vec<Order>>();`
`        orders`
`            .iter()`
`            .filter(|review| orders.customer_id == self.id)`
`            .collect()`
`    }`
`}`

and added the following above pub struct Customer 
`#[graphql(complex)]`

That didn't seem to help...

**Reply by cchee** [reply]
Can anyone please enlighten me? Or give me some pointers? Even comment to stay whether I am on the right track of getting close to make this work?

**Reply by cchee** [reply]
@sunli829, do you have some spare time to kindly shred some light to enlighten an apparently lost soul?

**Reply by cchee** [reply]
@sunli829 老油條, 可以請您幫個忙？請您多多指教

**Comment by sunli829** [comment, marked as answer]
See if this helps you? 

https://romankudryashov.com/blog/2020/12/graphql-rust/#_apollo_federation_support

**Reply by cchee** [reply]
Thank you @sunli829. Let me give that a try.

**Reply by cchee** [reply]
I am so excited. Nested query is working now!!! Thank you so much @sunli829!!!

**Reply by cchee** [reply]
@sunli829 By the way, it is working when I use multiple Apollo rover dev for subgraph services, but it couldn't resolve when using Apollo router with router.yaml. Is there something I missed in the router.yaml?

**Reply by cchee** [reply]
Figured it out... I need to re-publish the updated graphql schema file for all the subgraphs and the supergraph. After that, everything works.
